source/compiler/tests/__addressof.pwn

```diff
@@ -0,0 +1,48 @@
+#include <console>
+
+const func1_addr = __addressof(Func1); // error
```

Comment:
> Functions can otherwise be used before declaration.

This is because in order to call functions the compiler outputs `call .funcname`, so it could look up the function and obtain its address later, when assembling the code. But `__addressof` works as an expression, and expressions are handled at the first two compilation stages (parsing), and the compiler can't know a function/label address until it parses its definition at the 2'nd pass; it doesn't keep function addresses obtained at the 1'st pass since at the 2'nd one the generated code can change (e.g. because of `#if defined <func>` being used for conditional compilation). I think the only solution for this is to add an ability to force a "second" second pass, but this sounds more like a work for a separate PR, IMO.

Reply:
You could build up a list of unresolved addressof calls, just as `CONST.pri` locations (or however it is actually implemented), then iterate through them once the function is resolved.  A third compiler pass would very probably break a lot of stuff.

Reply:
Or, as I said, would `forward` fix this?  I somewhat doubt it as address isn't in the metadata that could be stored in that way.  But maybe the pending addresses could be added to the metadata for a function, thus require `forward` for this case as well (for something like consistency).

Reply:
I just realised this works:

```pawn
main()
{
    new a = __emit(CONST.pri Func);
}

Func()
{
}
```

Reply:
> I just realised this works:
` new a = __emit(CONST.pri Func);`

Indeed. As I already mentioned, `__emit` implements its own code generation mechanism, so it isn't bound by the limits of the standard codegen.